src/queries/role.ts

```diff
@@ -28,22 +31,90 @@ const matchAccessibleNameIfNeeded = (
   );
 };
 
+// disabled:undefined is equivalent to disabled:false, same for selected. busy not, but it makes
+// sense from a testing/voice-over perspective. checked and expanded do behave differently
+const implicityFalseState: AccessibilityStateField[] = [
+  'disabled',
+  'selected',
+  'busy',
+];
+
+const matchAccessibleStateIfNeeded = (
+  node: ReactTestInstance,
+  options?: ByRoleOptions
+) =>
+  accessibilityStates.every((accessibilityState) => {
+    const queriedState = options?.[accessibilityState];
+
+    if (typeof queriedState !== 'undefined') {
+      // Some accessibilityState properties have implicit value (when not set)
+      const defaultState = implicityFalseState.includes(accessibilityState)
+        ? false
+        : undefined;
+      return (
+        queriedState ===
+        (node.props.accessibilityState?.[accessibilityState] ?? defaultState)
+      );
+    } else {
+      return true;
+    }
+  });
+
+const accessibilityStates: AccessibilityStateField[] = [
+  'disabled',
+  'selected',
+  'checked',
+  'busy',
+  'expanded',
+];
+
 const queryAllByRole = (
   instance: ReactTestInstance
 ): ((role: TextMatch, options?: ByRoleOptions) => Array<ReactTestInstance>) =>
   function queryAllByRoleFn(role, options) {
-    return instance.findAll(
-      (node) =>
+    return instance.findAll((node) => {
+      // run the cheapest checks first, and early exit too avoid unneeded computations
```

Comment:
style: we do not need to put the early returns by hand. JS is smart enough to skil calculating `b()` in `a() && b()` when it knows that `a() == false`

```suggestion
    return instance.findAll((node) => {
      typeof node.type === 'string' &&
      matchStringProp(node.props.accessibilityRole, role) &&
      matchAccessibleStateIfNeeded(node, options) &&
      matchAccessibleNameIfNeeded(node, options?.name) 
```

| The logical AND expression is a short-circuit operator. As each operand is converted to a boolean, if the result of one conversion is found to be false, the AND operator stops and returns the original value of that falsy operand; it does not evaluate any of the remaining operands.

Consider the pseudocode below.

(some falsy expression) && expr
The expr part is never evaluated because the first operand (some falsy expression) is evaluated as [falsy](https://developer.mozilla.org/en-US/docs/Glossary/Falsy). If expr is a function, the function is never called. See the example below:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Logical_AND#short-circuit_evaluation

Reply:
Yeah, I'm well aware, I think it's just a left over from when we didn't have helpers function for all cases. I do tend to find explicit return more readable in a lot of cases, but here it probably doesn't matter!